### DISCUSSION TITLE
RAWG

### DISCUSSION THREAD

**Discussion by sgamesdev** [Leo-Corporation/Gavilya#263] [Q&A]
How did you use the RAWG API in your app

**Comment by lpeyr** [comment]
The RAWG API is used to fetch data about games information, such as the cover image, ratings, achievements, etc... To get this information, Gavilya uses RestSharp to fetch the RAWG API. But before that, the user must associate the game to a RAWG (game) ID using a search dialog. The RAWG ID of the game is then stored in the `Games.gav` file. After associating the game to a RAWG ID, Gavilya can use the data from the API.
For example, here's the code that gets the ratings of a game from its RAWG ID:
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya/Classes/Global.cs#L361-L387

> **Note** Most of the code in this method was generated using Nightingale REST Client. I had to update it since I updated RestSharp to a version that contained breaking changes.

More RAWG-related methods are available in the [`Global.cs`](https://github.com/Leo-Corporation/Gavilya/blob/main/Gavilya/Classes/Global.cs) file.

You'll notice that the JSON returned by the API is deserialized to a `Game` object which contains all different properties about a game.
It's available in `Gavilya.SDK` which is a separated module of Gavilya that is used as a library.
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya.SDK/RAWG/Game.cs#L1-L156

To use the API, you need an API key that isn't publicly available in this repo. I have two `APIKeys` class: one that is public and empty, and one that is local, so the app can still compile.
Here's how it looks like:
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya/Classes/APIKeys.cs#L25-L35

The local one: (named `APIKeysLocal.cs`)
~~~ cs
public static partial class APIKeys
{
	static APIKeys()
	{
		RAWGAPIKey = "API_KEY_HERE";
	}
}
~~~

You can learn more about RestSharp [here](https://restsharp.dev/).
You can check the RAWG API documentation [here](https://api.rawg.io/docs/).

**Reply by sgamesdev** [reply]
Is there anyway to get the game genres , to convert the list to a string ?

**Reply by lpeyr** [reply]
Yes, I guess you could use the following code (if you know the game id):
~~~ cs
// Don't forget to include the SDK.
using Gavilya.SDK.RAWG;


var client = new RestClient(new Uri($"https://api.rawg.io/api/games/{id}")); // Create a REST Client 
var request = new RestRequest 
{ 
 	Method = Method.Get 
}; // Create a request 
request.AddQueryParameter("key", APIKeys.RAWGAPIKey); 
var response = await client.ExecuteAsync(request); // Execute the request and store the result 
  
var game = JsonSerializer.Deserialize<Game>(response.Content); // Deserialize the content of the reponse 
string s = ""; // The string

for (int i = 0; i < game.genres.Length; i++)
{
	s += $"{game.genres[i].name}\n"; // Convert the list to a string
}
~~~
For Minecraft `s` would be equal to:
~~~ txt
Action
Arcade
Simulation
Indie
Massively Multiplayer
~~~

**Reply by sgamesdev** [reply]
how did you do the achievement system

**Reply by lpeyr** [reply]
Achievements are a bit different from other requests to the RAWG API. Indeed, the URL where the request is sent is not the same as usual, and it also returns a different JSON object, which will be deserialized into the `AchievementsResult` class. This class contains a `List` of `Achievement` named `results`. The `Achievement` class contains different property:
- id
- name
- description
- image
- percent (The percentage of players who have unlocked this achievement)

To get the achievements of a game, Gavilya uses the `GetAchievementsAsync()` method:
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya/Classes/Global.cs#L417-L443

> **Note** Gavilya only displays the first 20 achievements of a game.

Achievements are displayed in a section of the `GameInfoPage` using the `AchievementItem` user control, using the data that was returned by the RAWG API.
#### XAML
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya/UserControls/AchievementItem.xaml#L1-L17

#### C#
https://github.com/Leo-Corporation/Gavilya/blob/5aa610f7a1ee262ed266754bfd1efdfe46e2aba0/Gavilya/UserControls/AchievementItem.xaml.cs#L24-L56
